dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+        if (!activateAppRequest.ParseFromString(pdu.data()))
+        {
+            errorMessage = "Failed to parse ActivateAppRequest message";
+            RETURN_WIN32(ERROR_INVALID_DATA);
+        }
+
+        if (!IsActivationKindSupported(activateAppRequest.activation_kind()))
+        {
+            errorMessage = "Unsupported activation kind: " + std::to_string(activateAppRequest.activation_kind());
+            RETURN_WIN32(ERROR_NOT_SUPPORTED);
+        }
+
+        {
+            // Acquire lock before adding the activity Id to the m_activityMap but do not hold the lock while activating the app as 
+            // that can take some time and we do not want one app activation to block others.
+            auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), g_connectionManager.m_closing);
+
+            auto inserted{ g_connectionManager.m_activityMap.emplace(pdu.activity_id(), AppActivationInfo(pdu.activity_id())) };
+            if (!inserted.second)
+            {
+                errorMessage = "Activity ID of the new activate app request is already used for a previous request. Each new request must use a new activity ID.";
+                RETURN_WIN32(ERROR_ALREADY_EXISTS);
+            }
+        }
+
+        wil::com_ptr<IApplicationActivationManager> aam;
+        HRESULT hrCoCreateInstance{ CoCreateInstance(CLSID_ApplicationActivationManager, nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&aam)) };
+        if (FAILED(hrCoCreateInstance))
+        {
+            errorMessage = "Failed to CoCreateInstance CLSID_ApplicationActivationManager.";
+            RETURN_HR(hrCoCreateInstance);
+        }
+
+        std::wstring appUserModelId{ ::Microsoft::Utf8::ToUtf16(activateAppRequest.app_user_model_id()) };
+        DWORD processId{};
+
+        switch (activateAppRequest.activation_kind())
+        {
+            case Dvc::ActivationKind::Launch:
+                {
+                    std::wstring launchArgs;
+                    if (!activateAppRequest.arguments().empty())
+                    {
+                        Dvc::LaunchActivationArgs args;
+                        if (!args.ParseFromString(activateAppRequest.arguments()))
+                        {
+                            errorMessage = "Failed to parse launch arguments";
+                            RETURN_WIN32(ERROR_INVALID_DATA);
+                        }
+
+                        launchArgs = ::Microsoft::Utf8::ToUtf16(args.arguments());
+                    }
+                    
+                    HRESULT hrActivateApp{ aam->ActivateApplication(appUserModelId.c_str(), launchArgs.c_str(), AO_NONE, &processId) };
+                    if (FAILED(hrActivateApp))
+                    {
+                        errorMessage = "Failed to Activate app.";
+                        RETURN_HR(hrActivateApp);
+                    }
```

Comment:
This error message will be sent back to client side so the client has rich error info. Otherwise, it gets an error code with no idea when exactly went wrong during the complex process of launching the remote app.